#' @description: peak annotation with geneHancer

# this script will create a list of regions 50kb upstream and downstream of cell-type-specific DAR
# generated by the find_atac_dar.R script. The regions are for querying the GeneHancer database to identify
# chromatin interactions present in that area to compare to the CCAN calculated by Cicero
library(openxlsx)
library(dplyr)
library(biomaRt)
library(tibble)
setwd("/data/active_data/lzl/RenalTumor-20200713/DataAnalysis-20210803/scATAC")

# load DAR and filter by adjusted pval, distance from DAR to nearest gene, average log-fold change of DAR,
# and the proportion of cells containing the DAR
list.clusterID <- getSheetNames("4.Peak/celltype.sig.pos.DARs.xlsx")

list.genes_to_query <- sapply(list.clusterID, function(x){
    dar <- read.xlsx("4.Peak/celltype.sig.pos.DARs.xlsx",
                   sheet = x, rowNames = F) %>%
    rownames_to_column(var = "coord") %>%
    dplyr::filter(p_val_adj < 0.05 & avg_log2FC > 0.25) %>%
    dplyr::filter(distance < 50000) %>%
    dplyr::distinct(gene, .keep_all = TRUE) %>% # remove second occurrences of each gene
    dplyr::top_n(1000, pi) %>% # take top 1000 genes by pi value
    dplyr::distinct(gene)
})

# function for returning a bed file for each gene with the desired coordinates
PullRegionBed <- function(genes, extend, human) {
  seqlevels <- c(seq(1,24),"X","Y")
  bm <- getBM(attributes = c("hgnc_symbol","chromosome_name", "start_position", "end_position","ensembl_gene_id"),
              filters = "hgnc_symbol", values = genes, mart = human) %>%
              dplyr::filter(chromosome_name %in% seqlevels)
  regionStart <- bm$start_position - extend
  if(regionStart < 0) { # make sure coordinates are not negative at beginning of chrom
    regionStart = 1
  }
  regionEnd <- bm$end_position + extend # coordinates cannot extend past max of chrom
  chr <- paste0("chr", bm$chromosome_name)
  bed <- data.frame(bm$hgnc_symbol, chr, regionStart, regionEnd)
  return(bed)
}

# create a list of bed files to query corresponding to each cell type in the snATAC dataset
human <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
index <- seq(list.genes_to_query)
list.regions_to_query.bed <- lapply(index, function(x) {
  genes <- list.genes_to_query[[x]]
  PullRegionBed(genes, 50000, human)
})
names(list.regions_to_query.bed) <- list.clusterID
saveRDS(list.regions_to_query.bed, file = "4.Peak/geneHancer/DAR.nearestGene.extend50KB.rds")
write.xlsx(list.regions_to_query.bed, file = "4.Peak/geneHancer/DAR.nearestGene.extend50KB.xlsx", sheetName = list.clusterID, rowNames = F)


# write the bed files
res <- lapply(index, function(x) {
  bed <- list.regions_to_query.bed[[x]][2:4] #增加了一列gene name
  file <- paste0("4.Peak/geneHancer/GeneHancerQuery_",list.clusterID[[x]],".bed")
  write.table(bed, file = file, row.names = FALSE, quote = FALSE, col.names = FALSE)
})

# bed files have to be manually uploaded to query the ucsc genome table browser located at
# https://genome.ucsc.edu/cgi-bin/hgTables
# group: Regulation
# track: GeneHancer
# table: GH Interactions (DE) (geneHancerInteractionsDoubleElite) ##OR## GH Interactions (geneHancerInteractions)
# output format: all fields from selected table
# output file: "GeneHancer_de_conns_clusterID.txt"  ##OR## "GeneHancer_all_conns_clusterID.txt"

